src/python/pants/source/filespec.py

```diff
@@ -16,7 +16,7 @@ def glob_to_regex(pattern):
   :returns: A regex string that matches same paths as the input glob does.
   """
   out = ['^']
-  components = pattern.strip('/').replace('.', '[.]').split('/')
+  components = pattern.strip('/').replace('.', '[.]').replace('$', '[$]').split('/')
```

Comment:
I think it might make sense to consider a more general escape mechanism. This covers `$`, but there are other non-glob regex characters that can appear in paths.

Initially I was thinking we could just drop `re.escape` in here, but because we're actually trying to parse out the glob info, I don't think that would work.

Maybe on line: 37 https://github.com/pantsbuild/pants/pull/4350/files#diff-2d528bdfb8ee22a56a1b0660753b7a3cR37